File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.java
import java.util.*;
class Solution {
    public String[] solution(String[] files) {
        String[] answer = {};
        Arrays.sort(files,(a,b)->{
            String[] str1 = a.split("[0-9]");
            String[] str2 = b.split("[0-9]");
            String s1 = str1[0].toLowerCase();
            String s2 = str2[0].toLowerCase();
            if(s1.equals(s2)) {
                String sNum1 = a.substring(s1.length());
                String sNum2 = b.substring(s2.length());
                int num1 = getNumber(sNum1);
                int num2 = getNumber(sNum2);
                return num1 - num2;
            }
            return s1.compareTo(s2);
        });
        
        answer = files;
        return answer;
    }
    int getNumber(String s) {
        String str = "";
        for(char c : s.toCharArray()) {
            if(Character.isDigit(c) && str.length() <= 5) {
                str += c;
            }
            else {
                break;
            }
        }
        return Integer.parseInt(str);
    }
}

File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.js
function makeHead(str) {
    let head = '';
    let index = 0;
    for(let i = 0; i < str.length; i++) {
        if (str[i] >= '0' && str[i] <= '9') {
            index = i;
            break;
        }
        head += str[i];
    }
    return [head.toLowerCase(),index];
}

function makeNumber(str,index) {
    let count = 0;
    let number = '';
    for (let i = index; i < str.length; i++) {
        if (count === 5 || str[i] < '0' || str[i] > '9') break;
        if (str[i] >= '0' && str[i] <= '9') {
            count++;
            number += str[i];
        }
    }
    return Number(number);
}

function solution(files) {
    files.sort((a,b) => {
        const [h1,index1] = makeHead(a);
        const [h2,index2] = makeHead(b);
        const n1 = makeNumber(a,index1);
        const n2 = makeNumber(b,index2);
        if (h1 === h2) return n1 - n2;
        return h1.localeCompare(h2);
    });
    return files;
}